# Importing library & dataframe

In [1]:
# Import libraries & setup 'path'

import pandas as pd
import numpy as np
import os

path = r'/Users/peterkim/Document/sql/CareerFoundry/Immersion Prep/4. Python Fundamentals for Data Analysts/Instacart Basket Analysis'
path


'/Users/peterkim/Document/sql/CareerFoundry/Immersion Prep/4. Python Fundamentals for Data Analysts/Instacart Basket Analysis'

In [3]:
# Import dataframes

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
#df_ords_wr = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)


# 1. Consistency checks

## Data consistency checks

In [3]:
# Create a dataframe

df_test = pd.DataFrame()


In [4]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]


In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)


mix


In [7]:
# Convert datatype to 'str'

df_test['mix'] = df_test['mix'].astype('str')


## Finding missing values

In [8]:
df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]


In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:

df_prods.shape


(49693, 5)

In [12]:
# Remove null values in 'product_name'

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [13]:
df_prods_clean.shape


(49677, 5)

## Finding duplicates

In [14]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]


In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
# Remove duplicates in 'product_name'

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

# 2. Run the 'df.describe()' function on your 'df_ords' dataframe
## Interpret the output

In [19]:
# Frequency checking 

df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Answer: Minimum values, 0, for 'order_dow', 'order_hour_of_day', and 'days_since_prior_order' should be investigated further

# 3. Check for mixed-type data in your 'df_ords' dataframe

In [20]:
# Check for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


# 4. If you find mixed-type data, fix it.

### Answer: No mixed-type data to fix

# 5. Check for missing values
## Explain why you used your method of choice

In [21]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [22]:
df_ords.shape

(3421083, 7)

### Answer: There are 206,209 missing information in 'days_since_prior_order' column. This accounts for 6% out of all 3,421,083 orders. 

# 6. Address the missing values
## Explain why you used your method of choice

### Answer: I'm not using any method of choice to address the missing values because 6% of missing values in 'days_since_prior_order' column from all orders can be assumed enough to be new orders that couldn't have previous orders.

# 7. Check for duplicate values

In [23]:
# Checking for duplicate

df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

### Answer: No duplicate values

# 8. Address the duplicates.

### Answer: No duplicate values to address

# 9. Export your final, cleaned 'df_prods' and 'df_ords' data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [24]:
# Exporting 'products_clean.csv' and 'orders_clean.csv' in 'Prepared Data' folder

df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean.csv'))
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean.csv'))
